src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.cs

```diff
@@ -510,6 +510,43 @@ private void Get_CORINFO_SIG_INFO(MethodDesc method, CORINFO_SIG_INFO* sig, bool
             }
         }
 
+        private bool TryGetUnmanagedCallingConventionFromModOpt(MethodSignature signature, out CorInfoCallConv callConv)
+        {
+            callConv = CorInfoCallConv.CORINFO_CALLCONV_UNMANAGED;
+            if (!signature.HasEmbeddedSignatureData || signature.GetEmbeddedSignatureData() == null)
+                return false;
+
+            foreach (EmbeddedSignatureData data in signature.GetEmbeddedSignatureData())
+            {
+                if (data.kind != EmbeddedSignatureDataKind.OptionalCustomModifier)
+                    continue;
+
+                if (!(data.type is DefType defType))
+                    continue;
+
+                if (defType.Namespace != "System.Runtime.CompilerServices")
+                    continue;
+
+                switch (defType.Name)
+                {
+                    case "CallConvCdecl":
```

Comment:
There are comments elsewhere in this change that the "last" mod opt found is the one that "wins". This logic does not do that. Please make this consistent with the other logic.

Reply:
Based on chatting offline, our logic here is to take the first one in the metadata. Updated comments to describe it as such.

Reply:
Let's make sure we close the loop with @333fred about order of these modopts.

Reply:
The compiler, at least, is going to be more strict here: if we encounter a signature that has multiple modopts and the `CallKind` isn't `unmanaged`, we're just going to mark it as unsupported metadata. For emit, the plan is that they'll be emitted as the first modopts in metadata, right after the parameter index. I think that's what this code is doing, so I believe we're in agreement.

Reply:
@333fred That is largely correct. The second part of this is the order of multiple arguments. For example:

`delegate* unmanaged[CDecl, StdCall]<void> F`

We would like the have the "last one wins" semantics, which means the `modopt`s should be inserted into metadata in "reverse order". This matches the semantics of ILAsm and we would prefer to keep that if possible.

Reply:
I chatted with @333fred offline. I think I was attempting to be too clever here and it isn't going to work. We will always need to read all the `modopt`s. I don't think we need to worry about this for P8, but for RC we should update the implementation.

The reasoning is because from Roslyn's perspective `delegate* unmanaged[CDecl, StdCall]<void> F` is the same as `delegate* unmanaged[StdCall, CDecl]<void> G`. However, if the runtime doesn't read all of them we would treat `F` differently than `G` and thus cause very difficult failures to root cause.

/cc @davidwrighton @elinor-fung

Reply:
That sounds like a good change. From a function pointer statement for the purpose of casting/calling that makes sense. 

Of course in metadata for the purpose of overloading a virtual method, the order *does* matter, as custom modifier order *is* part of the signature check. I think we can take that as a fix post preview 8.

Reply:
Filed issue https://github.com/dotnet/runtime/issues/39378.